### ISSUE TITLE
Add the new view of genres 

### ISSUE THREAD

**Issue by tesshucom** [tesshucom/jpsonic#1844]
Scan considerations. Related airsonic/airsonic#1046,  #1736.
Prerequisites: [v111.6.0](https://github.com/tesshucom/jpsonic/milestone/62), #2274

As mentioned in airsonic/airsonic#1046, we may see genres with no content and zero count.
That is related to the problem of what genre the album will be when Songs with different genres exist in the same album. 

<details>
<summary>Their own specification </summary>

The original reason for the genre confusion is that Subsonic offered their own specification of "genre per album". It's a kind of summary rule.
 - `Genre by album` is something that doesn't exist in ID3 tags.
   - The relationship between one song and Genre is 1:n. Since an album has multiple songs, the relationship between one album and Genre is 1:n. Therefore, when albums and genres are expressed in a 1:1 ratio, information is always missing. Anyone who knows ID3 should know this in theory. 
   - However, in the early days of ID3, genres were rarely used, and generally started out with a simple usage of assigning the same genre to all songs on an album. (It is still used in this way in some cases.)
 - A similar proprietary spec example is Linn Kazoo's `Artist by genres`.

The advantage is that it is simple and fast, the disadvantage is that it may not be intuitive to the user. It's an implementation that aims to be the fastest while providing limited functionality. Also, they were made at a time when advanced genre management had just begun and environmental performance was very low. So rather than a bug, it's "that features is that kind of specification".
</details>
<details>
<summary>Transition proposals </summary>

 - As mentioned in `airsonic/airsonic#1046`, we may see genres with no content and zero count. This can be resolved by simply filtering. But that only masks the problem, it doesn't fundamentally solve the spec.
 - The view presented by pwepwe973 is probably what users generally expect.
   - This is the dedign of view we commonly see today. Display the genre of the song obediently. Group songs by album for easier understanding.

___

What is planned for Jpsonic is to add the new view of genres.

 - Restructuring to provide the views users expect within the bounds of the Subsonic API... This is technically possible, but will be considered later.
 - Keep the Subsonic API specification as current implementation. and provides a generic, Song-based album view.
 - Allow users to switch as a transitional measure

The expected phases are:

 - With UPnP, we can do it right away relatively. Because it's a view and model spec that can accommodate this out of the box. For a soft landing, we should aim there first.
 - We can then port that view to your web page.
 - Subsonic API will be considered later.
 - It's not finalized, but if the speed isn't a problem, we can consider removing the old spec and moving to the new spec entirely.

These are realistic transition proposals that consider conservatism and risk.

</details>

_____

**Comment by tesshucom** [comment]
There are two key points to make the genre filter function work:

 - 1. How to retain Genre Master. This is confirmed and will be handled on the Lucene side. It is not practical to do this on the database side. Specifically, there is no intuitive Master in the persistence layer; searches are performed on an ongoing basis. This is because the Music Folder filter exists before the Genre filter, making it a compound condition. (This is something that Subsonic and Airsonic do not take into account. However, this is the cause of the so-called "empty genre" problem, so it needs to be overcome.) Although it seems more complicated, this method is better in terms of fault tolerance. 
   - Depending on processing speed, a short-term cache may be used. (In other words, the Genre Master exists in memory until it expires or the Music Folder is updated.) If Lucene is faster than expected and memory consumption is limited, caching may be forgone. Expected speed is within 400ms/40k Songs. If it exceeds this limit significantly, some action will be required. 
 - 2. Changes to ID3Album update logic. 
   - The so-called ID3 Album is a user-friendly and traditional representation of the Artist/Album Group of the same name. The current method of differential updates is a tag base stored in a file. 
     - The method for obtaining differences during scanning is as follows: The reference file is determined among the child elements of Album, that is, the song data, and the changes is detected by comparing the common fields of "the Song record in the File Structure " and "the ID3 Album record". 
   - The Genre of ID3Album refers to the Genre tag set in the base file (multiple genres are already supported). This needs to be modified to reference the Genre of all child elements (songs) that exist within the Album.
     - This will make the previous differential comparison method unusable. In short, it is necessary to detect when the Genre of any child song has changed. (In reality, it is faster to detect the change in the child rather than limiting the Field.)

The latter logic is a bit more complex, but if well designed it could potentially be faster than the previous method. (For example, if we don't compare all fields, but just mention the date...)

**Comment by tesshucom** [comment]
**Changes to ID3Album update logic** will need to be addressed first.

As we can see from the scan log, Jpsonic has a different processing order than Airsonic/Subsonic. In particular, the analysis order of Album and Song is reversed. (That makes sense: to do this properly, we first need to analyze the song first among all the models.)

 - When a Song is updated, the children_last_updated of the Album in the File Structure (i.e., the parent directory) is set to a fixed value, FAR_FUTURE.
 - In a later process, Album records with children_last_updated set to FAR_FUTURE are detected and analyzed. If the update is successful, children_last_updated will be updated to the Scan Date.

Although it is redundant, the purpose is as follows:

 - No updates are missed. It is a common bug in media servers that the album information is not updated even if the file is changed.
 - Increase the certainty of data recovery. This is primarily intended for data recovery after cancellation (or catastrophic failure). If the problem is minor and the user notices it, scanning again should fix it. If the analysis is cancelled between the song analysis phase and the Album analysis phase, the Album analysis phase will be run correctly on the next scan, so to speak, "continuing from where it left off".

At the moment, the File Structure Album differential update works fine. There are no plans to change these designs.

It is necessary to change the ID3 Album update logic.

____

[hypothesis]

 - Currently, changes in Album ID3 records are detected by comparing the contents of the original song record and the Album ID3 record. Change this to date-based flag detection, as is done with Album in File Structure.
 - Currently, the children_last_updated field of a song record is unused and is filled with a fixed value, FAR_PAST. By entering FAR_FUTURE into children_last_updated only when song data is updated, changes can be detected using logic similar to that of albums in the file structure.
   - For the Album in the File Structure, if the Album record's children_last_updated is FAR_FUTURE, this indicates that the data needs to be refreshed in a later process.
   - For ID3 albums, if any of the song records have children_last_updated set to FAR_FUTURE, this indicates that the data needs to be updated at a later process.
 - In a later step, if the album update process (either File Structure or ID3) is successful, the flag will be updated. FAR_FUTURE is replaced by Scan Date.

[Expected impact]

 - The state transitions will be clearer. When we look at children_last_updated, we can see the unprocessed data (easier to debug).
 - As a result, if the Album is updated successfully, the Scan Date is inserted into children_last_updated. This indicates the normal update date of the so-called Jpsonic internal data.(last_scanned has a slightly different meaning. It just stamps the date of scanning.)
 - Performance impact
   - Subsequent scans (but not the first) will detect ID3 album changes sooner if there are no changes. This means that scanning may be faster. (To be verified)
   - If a large amount of song data tag changes have been made, updates will be slower. (It won't be significantly slower, but it will add a process to update the dates of child songs when updating an Album.)

____

What should be set to Genre in AlbumID3? Not clearly defined in the Subsonic API.

 - From the next version onwards, Jpsonic will use the concatenated string of child Song Genres. Simply concatenate distinct genres with a semicolon. There are various opinions on this, but it is assumed that there will be no impact on Client Apps (although there may be an impact on display. Rather, this problem should have already occurred when Multi Genre was introduced, so it is not a problem that is occurring now.).
   - At this stage, it is a simple character concatenation. This means that there may be some overlap in the case of multiple-genres.
     - There are a few ways of thinking about this. for example··· for example···
       - Eliminate duplicates on the apps side? No, there are no issues at the moment, so at least in the initial phase we have no plans to add any additional processing. This is a process that comes in exchange for performance.
       - In reality, this overlap does not need to be eliminated. The idea is that if genres are managed properly, then distinct child songs of an album can be concatenated without duplication.
     - Since it doesn't affect processing, it's best to make gradual adjustments while keeping in mind some considerations. 
   - The semicolon is used for this purpose and is fixed.
     - Because there is less risk. Customizing the delimiter is not common. (Even the ID3 standard separators can cause problems depending on the database. This is a very sensitive topic.) If it is absolutely necessary, it should be done on the View side (on the Client, not on the Server).

**Comment by tesshucom** [comment]
We tested the draft implementation to check scan speed and found that the impact on speed appears to be limited.

 - Obviously, single registrations and updates will be a bit slower than before. However, due to the change in the method of difference detection, when evaluated on a scanning procedure basis, it does not lead to a significant increase in time. When evaluated in a broader scope, the speed reduction and speed improvement cancel each other out to some extent.
 - "Repeated scans with no changes" will be now several seconds faster than before. The difference method for REFRESH_ALBUM_ID3 has been changed, so that the determination of no changes is made more quickly.
 - Scanning with IgnoreTimestamp enabled is faster than a completely new scan.

Generally, there appears to be no problem. To give an example of a use case, it would look something like this.

 - [A little bad] A complete initial scan will be a bit slower.
 - [Good] After the initial scan, the user adds several albums periodically. In this case, subsequent scans will be faster than before.
 - [Good] Occasionally a Jpsonic update may force you to do a full scan. This will be a bit faster than before.

The "A little bad" part can be sped up as much as you want later. Accurate processing is needed before any real speed improvements can be made. Also, although it will be a little slower, I think it will be close to the margin of error in the realistic range of less than 100,000 songs... In reality, the impact of hardware specs is much greater 🙄
